pyne/material_library.pyx

```diff
@@ -0,0 +1,332 @@
+"""Python wrapper for material library."""
+from __future__ import division, unicode_literals
+
+# Cython imports
+from libcpp.utility cimport pair as cpp_pair
+from libcpp.set cimport set as cpp_set
+from libcpp.string cimport string as std_string
+from libc.stdlib cimport malloc, free
+from libcpp.unordered_map cimport unordered_map as cpp_umap
+from libcpp.vector cimport vector as cpp_vector
+from libcpp cimport bool as cpp_bool
+# from cython.operator cimport reference as ref
+from cython.operator cimport dereference as deref
+from cython.operator cimport preincrement as inc
+
+# Python imports
+import collections
+cimport numpy as np
+import numpy as np
+from warnings import warn
+from pyne.utils import QAWarning
+import os
+import sys
+if sys.version_info[0] >= 3:
+    # Python2 basestring is now Python3 string
+    basestring = str
+
+import tables as tb
+
+# local imports
+from pyne cimport cpp_material
+from pyne cimport cpp_material_library
+cimport pyne.stlcontainers as conv
+import pyne.stlcontainers as conv
+
+cimport cpp_jsoncpp
+cimport pyne.jsoncpp as jsoncpp
+import pyne.jsoncpp as jsoncpp
+
+cimport pyne.nucname as nucname
+import pyne.nucname as nucname
+
+cimport pyne.data as data
+import pyne.data as data
+
+cimport pyne.material as material
+import pyne.material as material
+
+
+warn(__name__ + " is not yet QA compliant.", QAWarning)
+
+# Maximum 32-bit signed int
+DEF INT_MAX = 2147483647
+
+_INTEGRAL_TYPES = (int, np.integer, np.bool_)
+ 
+
+#
+#  Material Library
+#
+
+cdef class _MaterialLibrary:
+    """This class allows the definitions of a set of material, stored by names
+    """
+
+
+    def __cinit__(self, lib=None, datapath="/materials", nucpath="/nucid"):
+        """Parameters
+        ----------
+        lib : dict-like, str, or None, optional
+            The data to intialize the material library with.  If this is a
+            string, it is interpreted as a path to a file.
+        datapath : str, optional
+            The path in the heirarchy to the data table in an HDF5 file.
+        nucpath : str, optional
+            The path in the heirarchy to the nuclide array in an HDF5 file.
+
+        """
+        if lib != None:
+            if sys.version_info[0] >= 3 and isinstance(lib, bytes):
+                lib = lib.decode()
+            if isinstance(lib, collections.Mapping):
+                self._inst = new cpp_material_library.MaterialLibrary()
+                list_ = []
+                for key in sorted(lib.keys()):
+                    mat = lib[key]
+                    self.__setitem__(key, material.ensure_material(mat))
+            elif isinstance(lib, basestring):
```

Comment:
Suggestion: Add some comments here for each case, for better understanding. Point out this block work with python2/3 as `basestring` are interpreted into different things in python2/3.